src/duk_api_debug.c

```diff
@@ -140,22 +142,53 @@ DUK_EXTERNAL void duk_debugger_cooperate(duk_context *ctx) {
 	DUK_UNREF(processed_messages);
 }
 
+DUK_EXTERNAL duk_bool_t duk_debugger_notify(duk_context *ctx, duk_idx_t nvalues) {
+	duk_hthread *thr;
+	duk_idx_t top, idx;
+	duk_bool_t ret = 0;
+
+	DUK_ASSERT_CTX_VALID(ctx);
+	thr = (duk_hthread *) ctx;
+	DUK_ASSERT(thr != NULL);
+	DUK_ASSERT(thr->heap != NULL);
+
+	if (DUK_HEAP_IS_DEBUGGER_ATTACHED(thr->heap)) {
+		top = duk_get_top(ctx);
+		if (top < nvalues) {
+			DUK_ERROR_API((duk_hthread *) ctx, "not enough stack values for notify");
+		}
+		duk_debug_write_notify(thr, DUK_DBG_CMD_APPNOTIFY);
+		for (idx = top - nvalues; idx < top; idx++) {
+			duk_tval *tv = duk_get_tval(ctx, idx);
```

Comment:
Since we're in trusted code and the top has been validated, you can use `DUK_GET_TVAL_POSIDX(ctx, idx)` here. It will translate to `thr->valstack_bottom + (idx)` without validation, which is safe as long as we know the top hasn't changed.